bp/src/BPSimulator.m

```diff
@@ -36,22 +39,50 @@ @implementation BPSimulator
 + (instancetype)simulatorWithConfiguration:(BPConfiguration *)config {
     BPSimulator *sim = [[self alloc] init];
     sim.config = config;
+    sim.testHostSimTemplates = [[NSMutableDictionary alloc] init];
+    sim.simDevices = [[NSMutableDictionary alloc] init];
     return sim;
 }
 
-- (NSMutableDictionary *)createSimulatorAndInstallAppWithBundles:(NSArray<BPXCTestFile *>*)testBundles {
-    NSMutableDictionary* testHostSimTemplates = [[NSMutableDictionary alloc] init];
-    NSString *simulatorUDIDString = nil;
+- (NSMutableDictionary *)createSimulatorAndInstallAppWithBundles:(NSArray<BPXCTestFile *>*)testBundles withSimTemplate:(NSString *)simTemplateUDID {
     NSError *error = nil;
+    NSString *simulatorUDIDString = nil;
+    SimServiceContext *sc = [SimServiceContext sharedServiceContextForDeveloperDir:self.config.xcodePath error:&error];
+    if (!sc && error) {
+        [BPUtils printInfo:ERROR withString:@"SimServiceContext failed: %@", [error localizedDescription]];
+        return nil;
+    }
+    SimDeviceSet *deviceSet = [sc defaultDeviceSetWithError:&error];
+    if (!deviceSet && error) {
+        [BPUtils printInfo:ERROR withString:@"SimDeviceSet failed: %@", [error localizedDescription]];
+        return nil;
+    }
     if (self.config.appBundlePath) {
-        // This is for integration testing for bluepill and bluepill-cli when we assign self.config.appBundlePath
-        simulatorUDIDString = [self installApplicationWithHost:self.config.appBundlePath withError:&error];
+        // This is for integration testing for bp and bluepill when we assign self.config.appBundlePath
+        SimDevice *simDevice = nil;
+        if (simTemplateUDID) {
+            simulatorUDIDString = simTemplateUDID;
+            simDevice = self.simDevices[simulatorUDIDString];
+        } else {
+            simDevice = [self createSimTemplateWithDeviceSet:deviceSet errPtr:&error];
+            self.simDevices[simDevice.UDID.UUIDString] = simDevice;
+        }
+        if (!simDevice || error) {
+            [BPUtils printInfo:ERROR withString:@"Create of simulator with error: %@", error];
+            return nil;
+        }
+        [simDevice bootWithOptions:nil error:&error];
+        if (error) {
+            [BPUtils printInfo:ERROR withString:@"Boot simulator failed with error: %@", [error localizedDescription]];
+            return nil;
+        }
+        simulatorUDIDString = [self installApplicationWithHost:self.config.appBundlePath withDeviceSet:deviceSet withSimDevice:simDevice withError:&error];
```

Comment:
let's rename this method to `installApplication:withDeviceSet:withSimDevice:withError`

Reply:
Thanks for the suggested changes. Will merge them.